README.md
erepAPI
=======

erepublik.com API wrapper.

Official erepublik API "Documentation" http://api.erepublik.com/doc/

"map","industries" and "war" are not currently supported due to issues with official API at the time of writing.

# Installation

Download this repository and modify `public_key` and `private_key` in `erepublik.py` (lines 34 and 35) to match your own personal keys.

You can then just import this module directly from directory or run `setup.py install` to install it to your `PATH`.

# Classes

## Citizen

Requires player ID (as string or integer) as initial argument. Raises `invalidID` exception if not supplied or invalid ID

### Example usage


    >>> from erepublik import Citizen
    >>> player=Citizen("1649964") # where "1649964" is player ID
    >>> player
    <erepublik.Citizen object at 0xb743f14c>
    >>> player.national_rank
    202
    >>> player.strength
    20822.371
    >>> player.achievements["Total"]
    182
    >>> player.achievements["President"]
    1
    >>> player.in_unit
    True
    >>> for key,value in player.achievements.items():
    ...     print key,value
    ...
    Ambassador 1
    Hero 21
    Super trooper 83
    Hard worker 44
    Mercenary 1
    Congressman 6
    Media mogul 2
    True patriot 19
    Resistance hero 2
    Campaign hero 1
    President 1
    Total 182
    Society builder 1
    >>>



### Full list of player variables available in Citizen class

| Variable        | Description           | Type  |
|----------------|:---------------------:|:-----:|
| `id`        | supplied user ID         | string |
| `name`        | Profile name              |   string |
|`is_alive` | Is user profile alive |boolean|
|`has_avatar` | Whether user has custom avatar or not|boolean|
|`experience_points`|Current number of experience points user has|integer|
|`level`|Current level number|integer|
|`birthday`|Date of profile creation, string. `MMM DD, YYYY`|string|
|`national_rank`|Current national rank number|integer|
|`profile_url`|full URL to user profile based on ID|string|
|`strength`|current number of strength points|float|
|`rank_points`|number of rank points|integer|
|`rank_stars`|number of rank stars|integer|
|`rank_icon`| URL to rank icon file|string|
|`rank_value`|Rank value between lowest(0) and highest(65) rank|integer|
|`achievements`|Dictionary of achievements with keys as achievement name and quantity(`integer`) as value.|Dictionary|
|`citizenship_country_id`|self explanatory |integer|
|`citizenship_country_name`|self explanatory| string|
|`citizenship_country_initials`|self explanatory|string|
|`citizenship_region_id`|self explanatory|integer|
|`citizenship_region_name`|self explanatory|string|
|`residence_country_id`|self explanatory |integer|
|`residence_country_name`|self explanatory|string|
|`residence_country_initials`|self explanatory |string|
|`residence_region_id`|self explanatory|integer|
|`residence_region_name`|self explanatory|string|
|`party_member`| true if player is part of political party otherwise false.|boolean|
||if player is member of political party, the following attributes are also available:|||
|`party_id`| id of the party|integer|
|`party_name`| name of the party |string|
|`is_president`| `True` if selected player is party president, otherwise `False`|boolean|
|`in_unit`| true if player is part of political party otherwise false| boolean|
||If player is member of military unit the following attributes are also available:||
|`unit_id`| id of the military unit|integer|
|`unit_name`|name of the unit|string|
|`unit_leader`|true if player is leader of the military unit, otherwise false|boolean|
|`owns_newspaper`|True if player owns newspapers otherwise false|boolean|
||If player owns newspapers the following attributes are also available||
|`newspaper_id`| id of the newspapers|integer|
|`newspaper_name`|name of the newspapers|string|


## Country_regions

Requires country ID (as `string` or `integer`) as initial argument.

### Example usage

    >>> from erepublik import Country_regions
    >>> regions=Country_regions(65).regions
    >>> for k,v in regions.items():
    ...     print k,v["id"]
    ...
    Eastern Serbia 637
    Belgrade 635
    Vojvodina 634
    Kosovo 743
    Southern Serbia 640
    Western Serbia 638
    Sumadija 636
    Raska 639

Each key in `regions` is `region name`(string) and each corresponding value is dict and contains following key, value pairs:

| Key        | Value Description           |  Value Type  |
|----------------|:---------------------:|:-----:|
|`id`|region id|integer|
|`owner_id`|ID of the country that is currently owner of the region|integer|
|`original_owner_id`|ID of the country that was/is original owner|integer|
|`url`|full url to ingame page with region info|string|


## Countries

No initial parameters required.

Used to list all countries or to find country by name or by an ID and list its information.

Each item in returned dictionary is `string` only `continet_name` is `None`

### Example usage

    >>> for country in Countries().all_countries:
    ...     print country["id"],country["initials"]
    ...
    167 AL
    27 AR
    50 AU
    33 AT
    83 BY
    32 BE
    76 BO
    [...]

    >>> print Countries().by_id(65)
    {
        u'name': u'Serbia',
        u'color': u'FFB47F',
        u'capital_region_name': u'Belgrade',
        u'capital_region_id': u'635',
        u'continent_id': u'1',
        u'continent_name': None,
        u'id': u'65',
        u'initials': u'CS'
    }
    >>> print Countries().by_name("serbia")["initials"]
    CS

`all_countries` - is a `list` of dictionaries each with following structure:

    {
        "id":"string",
        "name":"string",
        "initials":"string",
        "color":"string",
        "continent_id":"string",
        "continent_name":None,
        "capital_region_id":None/string,
        "capital_region_name":None/string
    },

## Regions

Requires valid region ID and, optionally, page number as either `string`s or `integer`s; larger regions contain multiple pages of data.

Contains list of citizen IDs as integers who are living in that region.


### Example usage

    >>> from erepublik import Region
    >>> Region(450,page=2).citizenIDs
    [1379069, 1379214, 1380608, 1380647, 1382406, 1382685,... ]

## Battle

Requires valid battle ID (string/integer) as initial argument.

### Example usage

    >>> from erepublik import Battle
    >>> b=Battle(39841)
    >>> b.is_resistance
    False
    >>> b.region_id
    450
    >>> b.start
    '2013-03-29 08:55:12'
    >>> print b.end
    None
    >>> b.defender_id
    59
    >>> b.defender_initials
    u'TH'
    >>> for k,v in b.defender_alies.items():
    ...     print k,v["id"]
    ...
    Poland 35
    Indonesia 49
    Serbia 65
    Republic of Macedonia (FYROM) 79
    [...]

### Full list of battle variables available in Battle class

| Variable        | Description           | Type  |
|----------------|:---------------------:|:-----:|
|`region_id`|id of the on which the battle is being fought on|integer|
|`region_name`|name of the region|string|
|`start`|date and time of the battle start, `YYYY-MM-DD HH:MM:SS` format|string|
|`end`|date and time of the battle end `YYYY-MM-DD HH:MM:SS` format, otherwise `None`|string|
|`finish_reason`|reason for battle finish if the battle has ended|string|
|`defender_id`|id of the country that's being attacked(defender)|integer|
|`attacker_id`|id of the country that's attacking(attacker) defender|integer|
|`defender_initials`|initials of the country that's being attacked(defender)|string|
|`attacker_initials`|initials of the country that's attacking(attacker)|string|
|`defender_allies`|allies of the country that's being attacked(defender)|dictionary|
|`attacker_allies`|allies of the country that's attacking(attacker)|dictionary|

Allies dictionary, country names are strings:

{
    country_name:dictionary,
    country_name:dictionary,
    [...]
}


Allies dictionary values structure:

| Key        | Value Description           | Type  |
|----------------|:---------------------:|:-----:|
|`"id"`|Country ID of the ally country|integer|
|`"initials"`|Country initials of the ally country|string|

___

# Licence

Copyright (c) 2013 Nikola Kovacevic nikolak@outlook.com,nikola.kovacevic91@gmail.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2013-09-22)
++++++++++++++++++

* First release on PyPI.

erepublik.py
# Copyright (c) 2013 Nikola Kovacevic   <nikolak@outlook.com>,
#                                       <nikola.kovacevic91@gmail.com>

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE

import hmac
import json
from hashlib import sha256
from sys import version_info
from time import strftime, gmtime, localtime

try:
    # Python 2
    from urllib2 import urlopen, Request
except ImportError:
    # Python 3
    from urllib.request import urlopen, Request


api_url = "http://api.erepublik.com/"
public_key = "public_key"
private_key = "private_key"


def _construct_url(resource, action, params=None):
    """
    Constructs rest of API url based on resource, action and parameters.

    Multiple parameters must be a list.

    Single parameters must be a single string

    Internal usage
    """
    final_url = '{}/{}'.format(resource, action)
    if type(params) == list:
        final_url += '?'
        for i in range(len(params)):
            if i < len(params) - 1:
                final_url += params[i] + '&'
            else:
                final_url += params[i]
    elif type(params) == str:
        final_url += '?{}'.format(params)
    else:
        pass
    return final_url


def _construct_headers(url):
    """
    Constructs headers to send with request.
    Internal usage.

    The Date header needs to be the current date and
    time of the request in RFC1123 format (e.g. Tue, 04 Sep 2012 15:57:48)

    The Auth header structure is {public_key}/{digest} where {public_key}
    is your public key provided by eRepublik and {digest}
    is a hash that represents an encrypted value of a concatenated string
    formed from {resource}, {action}, {params} (if they exist),
    and {date} (the same value Date header has).

    """
    header = {"Auth": public_key + '/'}
    date = strftime("%a, %d %b %Y %H:%M:%S", gmtime())
    header["Date"] = date
    to_digest = (url.replace('/', ':').replace('?', ':') + ':').lower() + date
    if version_info >= (3, 0):
        # 3.X behavior is correct. Since HMAC operates on bytes, not text,
        # only bytes are accepted. In Python 2, the acceptance of
        # Unicode strings is more an accident than a feature.
        header["Auth"] += hmac.new(bytes(private_key, 'utf-8'),
                                   bytes(to_digest, 'utf-8'),
                                   sha256).hexdigest()
    else:
        header["Auth"] += hmac.new(private_key,
                                   to_digest,
                                   sha256).hexdigest()
    return header


def _load(url, headers):
    """
    Returns json/dictionary of data in "message" from API response
    """
    req = Request(api_url + url)
    req.add_header("Date", headers["Date"])
    req.add_header("Auth", headers["Auth"])
    response = urlopen(req)
    content = response.read().decode("utf-8")
    data = json.loads(content)
    try:
        if data["code"] != 200:
            raise invalidCode
        else:
            return data["message"]
    except:
        raise urlParseError


class invalidID(Exception):
    pass


class invalidCode(Exception):
    pass


class urlParseError(Exception):
    pass


class Citizen(object):

    """Contains all citizen variables
        requires valid citizen ID as initial parameter
    """
    def __init__(self, citizenID):
        super(Citizen, self).__init__()
        self.id = str(citizenID)
        self.resource = "citizen"
        self.action = "profile"
        self.params = "citizenId=" + self.id
        self.profile_url = "http://www.erepublik.com/en/citizen/profile/" + self.id

        if not self.id.isdigit():
            raise invalidID
        else:
            self.url = _construct_url(self.resource, self.action, self.params)
            self.headers = _construct_headers(self.url)
        self.data = _load(self.url, self.headers)

        # General attributes
        self.name = self.data["general"]["name"]
        if self.name is None:
            raise invalidID
        self.is_alive = self.data["general"]["is_alive"] == "1"
        self.has_avatar = self.data["general"]["has_avatar"]
        self.avatar = self.data["general"]["avatar"]
        self.experience_points = int(self.data["general"]["experience_points"])
        self.level = int(self.data["general"]["level"])
        self.birthday = self.data["general"]["birthDay"]
        self.national_rank = self.data["general"]["nationalRank"]

        # Military attributes
        self.strength = float(self.data["militaryAttributes"]["strength"])
        self.rank_points = int(self.data["militaryAttributes"]["rank_points"])
        self.rank_name = self.data["militaryAttributes"]["rank_name"]
        self.rank_stars = self.data["militaryAttributes"]["rank_stars"]
        self.rank_icon = self.data["militaryAttributes"]["rank_icon"]
        self.rank_value = self.__return_rank_value()

        # Achievements - dict
        if self.data["achievements"]:
            self.achievements = self.__parse_achievements()
        else:
            self.achievements = {}

        # Location
        self.citizenship_country_id =\
            int(self.data["location"]["citizenship_country_id"])
        self.citizenship_country_name =\
            self.data["location"]["citizenship_country_id"]
        self.citizenship_country_initials =\
            self.data["location"]["citizenship_country_initials"]
        self.citizenship_region_id =\
            int(self.data["location"]["citizenship_region_id"])
        self.citizenship_region_name =\
            self.data["location"]["citizenship_region_name"]
        self.residence_country_id =\
            int(self.data["location"]["residence_country_id"])
        self.residence_country_name =\
            self.data["location"]["residence_country_name"]
        self.residence_country_initials =\
            self.data["location"]["residence_country_initials"]
        self.residence_region_id =\
            int(self.data["location"]["residence_region_id"])
        self.residence_region_name =\
            self.data["location"]["residence_region_name"]

        # Party info
        if self.data["party"]:
            self.party_member = True
            self.party_id = int(self.data["party"]["id"])
            self.party_name = self.data["party"]["name"]
            self.is_president = self.data["party"]["is_president"] == "1"
        else:
            self.party_member = False

        # Military unit
        if self.data["militaryUnit"]:
            self.in_unit = True
            self.unit_id = int(self.data["militaryUnit"]["id"])
            self.unit_name = self.data["militaryUnit"]["name"]
            self.unit_leader = self.data["militaryUnit"]["is_leader"]
        else:
            self.in_unit = False

        # Newspappers
        if self.data["newspaper"]:
            self.owns_newspaper = True
            self.newspaper_id = int(self.data["newspaper"]["id"])
            self.newspaper_name = self.data["newspaper"]["name"]
        else:
            self.owns_newspaper = False

    def __parse_achievements(self):
        """
        Returns a dictionary with all medals/achievements and
        total number of them from selected citizen

        {
            "achievement_name":quantaty_integer,
        }
        """
        total_count = 0
        parsed = {}
        for item in self.data["achievements"]:
            ach_type = self.data["achievements"][item]["type"].capitalize()
            ach_count = int(self.data["achievements"][item]["total"])
            parsed[ach_type] = ach_count
            total_count += ach_count
        parsed["Total"] = total_count
        return parsed

    def __return_rank_value(self):
        full_rank_name = self.rank_name + "*" * self.rank_stars
        rank_list = {
            'Recruit': 1,
            'Private': 2,
            'Private*': 3,
            'Private**': 4,
            'Private***': 5,
            'Corporal': 6,
            'Corporal*': 7,
            'Corporal**': 8,
            'Corporal***': 9,
            'Sergeant': 10,
            'Sergeant*': 11,
            'Sergeant**': 12,
            'Sergeant***': 13,
            'Lieutenant': 14,
            'Lieutenant*': 15,
            'Lieutenant**': 16,
            'Lieutenant***': 17,
            'Captain': 18,
            'Captain*': 19,
            'Captain**': 20,
            'Captain***': 21,
            'Major': 22,
            'Major*': 23,
            'Major**': 24,
            'Major***': 25,
            'Commander': 26,
            'Commander *': 27,
            'Commander **': 28,
            'Commander ***': 29,
            'Lt Colonel': 30,
            'Lt Colonel*': 31,
            'Lt Colonel**': 32,
            'Lt Colonel***': 33,
            'Colonel': 34,
            'Colonel*': 35,
            'Colonel**': 36,
            'Colonel***': 37,
            'General': 38,
            'General*': 39,
            'General**': 40,
            'General***': 41,
            'Field Marshal': 42,
            'Field Marshal*': 43,
            'Field Marshal**': 44,
            'Field Marshal***': 45,
            'Supreme Marshal': 46,
            'Supreme Marshal*': 47,
            'Supreme Marshal**': 48,
            'Supreme Marshal***': 49,
            'National Force': 50,
            'National Force*': 51,
            'National Force**': 52,
            'National Force***': 53,
            'World Class': 54,
            'World Class': 55,
            'World Class': 56,
            'World Class': 57,
            'Legendary Force': 58,
            'Legendary Force*': 59,
            'Legendary Force**': 60,
            'Legendary Force***': 61,
            'God of War': 62,
            'God of War*': 63,
            'God of War**': 64,
            'God of War***': 65,
        }
        return rank_list.get(full_rank_name, None)


class Country_regions(object):

    """Contains regions of country as dictionary.
      Requires country ID as initial argument

    self.regions={ "region_name":{
                                    "id":integer
                                    "owner_id":integer
                                    "original_owner_id":integer
                                    "url":string
                                }

                }
    """
    def __init__(self, countryID):
        super(Country_regions, self).__init__()
        self.id = str(countryID)
        self.resource = "country"
        self.action = "regions"
        self.params = "countryId=" + self.id

        if not self.id.isdigit():
            raise invalidID
        else:
            self.url = _construct_url(self.resource, self.action, self.params)
            self.headers = _construct_headers(self.url)
        self.data = _load(self.url, self.headers)
        self.base_url = "http://www.erepublik.com/en/main/region/"
        self.regions = {}
        for item in self.data["regions"]["region"]:
            region = self.data["regions"]["region"][item]
            self.regions[region["name"]] = {
                "id": int(region["id"]),
                "owner_id": int(region["current_owner_country_id"]),
                "original_owner_id": int(region["original_owner_country_id"]),
                "url": self.base_url + region["permalink"]}


class Countries(object):

    """Contains list of countries
        Use `by_id(country_id)` to get basic info about country as dict
            {
               "id":string,
               "name":string,
               "initials":string,
               "color":string,
               "continent_id":string,
               "continent_name":null,
               "capital_region_id":null/string,
               "capital_region_name":null/string
            },
    """
    def __init__(self):
        super(Countries, self).__init__()
        self.resource = "countries"
        self.action = "index"
        self.url = _construct_url(self.resource, self.action)
        self.headers = _construct_headers(self.url)
        self.data = _load(self.url, self.headers)
        self.all_countries = self.data["countries"]["country"]

    def by_id(self, id):
        """
        Returns Country info as dict
        Requires valid country ID
        """
        country_data = None
        for item in self.all_countries:
            if item["id"] == str(id):
                country_data = item
                break
        return country_data

    def by_name(self, name):
        """
        Returns Country info as dict
        Requires valid country name
        """
        country_data = None

        for item in self.all_countries:
            if item["name"].lower() == name.lower():
                country_data = item
                break
        return country_data


class Region(object):

    """
    Contains list of citizen IDs living in that region in self.citizenIDs
    Requires valid region ID as input and optional page number
    otherwise 1 is used for page num
    """
    def __init__(self, regionID, page=1):
        super(Region, self).__init__()
        self.id = str(regionID)
        self.page = str(page)
        self.resource = "region"
        self.action = "citizens"
        self.params = ["regionId=" + self.id, "page=" + self.page]

        if not self.id.isdigit():
            raise invalidID
        else:
            self.url = _construct_url(self.resource, self.action, self.params)
            self.headers = _construct_headers(self.url)
        self.data = _load(self.url, self.headers)
        if self.data["citizens"] is False:
            self.citizenIDs = None
        else:
            self.citizenIDs = [int(cit_id["citizen_id"]) for cit_id in self.data["citizens"]]


class Battle(object):

    """docstring for Battle"""
    def __init__(self, battleID):
        super(Battle, self).__init__()
        self.id = str(battleID)
        self.resource = "battle"
        self.action = "index"
        self.params = "battleId=" + self.id
        if not self.id.isdigit():
            raise invalidID
        else:
            self.url = _construct_url(self.resource, self.action, self.params)
            self.headers = _construct_headers(self.url)
        self.data = _load(self.url, self.headers)

        self.is_resistance = self.data["battle"]["is_resistance"] == "true"

        # Region
        self.region_id = int(self.data["battle"]["region"]["id"])
        self.region_name = self.data["battle"]["region"]["name"]

        # Progress
        self.start = self._format_time(self.data["battle"]["progress"]["started-at"])
        self.end = self._format_time(self.data["battle"]["progress"]["finished-at"])
        self.finish_reason = self.data["battle"]["progress"]["finished-reason"]

        # Defenders
        self.defender_id = int(self.data["battle"]["progress"]["countries"]["victim_country"]["id"])
        self.defender_initials = self.data["battle"]["progress"]["countries"]["victim_country"]["initials"]
        self.defender_allies = {}
        for item in self.data["battle"]["progress"]["countries"]["victim_country"]["allies"]:
                country_name = item["country"]["name"]
                country_id = int(item["country"]["allied_country_id"])
                country_initials = item["country"]["initials"]
                self.defender_allies[country_name] = {"id": country_id,
                                                     "initials": country_initials
                                                     }

        # Invaders
        self.defender_id = int(self.data["battle"]["progress"]["countries"]["invader_country"]["id"])
        self.defender_initials = self.data["battle"]["progress"]["countries"]["invader_country"]["initials"]
        self.defender_allies = {}
        for item in self.data["battle"]["progress"]["countries"]["invader_country"]["allies"]:
                country_name = item["country"]["name"]
                country_id = int(item["country"]["allied_country_id"])
                country_initials = item["country"]["initials"]
                self.defender_allies[country_name] = {"id": country_id,
                                                     "initials": country_initials
                                                     }

    def _format_time(self, api_time):
        if api_time.isdigit() is False:
            return None
        else:
            return strftime('%Y-%m-%d %H:%M:%S', localtime(int(api_time)))
